### ISSUE TITLE
impl FromIterator<pest::iterators::Pair<'_, RuleType>> for Pairs

### ISSUE THREAD

**Issue by Redrield** [pest-parser/pest#238]
At the moment, there is an impl for `Iterator` on Pairs. This allows me to use std iterator combinators like filter and map. However there is no impl for `FromIterator<Pair<'_, RuleType>>` for Pairs, which means that after using the combinators I am unable to convert back to a Pairs<RuleType> to use from the rest of my code.

**Comment by dragostis** [comment]
This cannot be implemented with the current API since an `Iterator<Item=Token>` doesn't always convert to a `Pairs` since the tokens might not have pairs and `FromIterator` doesn't accept failure. One could implement a way to convert the tokens back into pairs, but the interior logic would have to be rewritten since `Pairs` currently use an efficient internal representation.

I do agree that this would be useful, but it's currently a bit out of reach for the near future. Do you have a particular use case in mind? Maybe there is a way to circumvent this particular need.

**Comment by Redrield** [comment]
I found a way around the use case that I had a problem with, but what I wanted to do was make it so that I could filter out comments from the pairs that I was given by my parser. The problem is that I had exported higher level functionality into different methods to sort of compartmentalize the way I turned the tree that pest gave me into one that I could use easier for code generation. So I tried to do `pairs.iter().filter(|pair| pair.as_rule() != Rule::comment).collect::<Pairs<Rule>>()`, and it didn't work because of the lack of an impl.
